index.rst

```diff
@@ -13,17 +13,256 @@
+``Butler.get``
+""""""""""""""
+
+1. If given a data ID, dataset type name, and collection search path instead of a ``DatasetRef``, obtain both the ``DatasetRef`` and all related datastore records from the database in a single Registry or butler server call.
+   If given a ``DatasetRef``, use this to obtain the datastore records. again via a single Registry or a butler server call.
+   ``QuantumBackedButler`` will look up datastore records directly in the quantum.
+
+2. Call ``Datastore.get`` with both the resolved ``DatasetRef`` and the bundle of records, returning the result to the caller.
+
+Because this is a read-only operation, consistency in the presence of failures is not a concern, but this still has a major advantage over the current approach for client-server in particular, as it bundles all http server access into a single call, followed by a direct object-store call, reducing latency and again allowing the Datastore to be a regular ``FileDatastore``.
+
+``Butler.unstore``
+""""""""""""""""""
+
+This is a proposed new interface for removing multiple datasets from the Datastore without removing them from the Registry - one part of a replacement for ``Butler.pruneDatasets``, and part of a reimplementation for ``Butler.removeRuns``.
+
+1. Pass the inputs to Registry and/or butler server to obtain ``DatasetRefs`` and datastore records, instructing it to delete those records at the same time.
+   ``QuantumBackedButler`` may not need to implement this operation at all, but if it does (e.g. for clobbering), it already has everything it needs in the quantum.
+   Deletion in the Registry can be made consistent via transactions, and in the client/server these can be started and committed entirely in the server.
+
+2. Pass the records to the Datastore and tell it to delete those artifacts.
+   Failures at this stage would not restore the Registry records for already-deleted datasets, leaving them in our undesirable-but-tolerable Datastore-only state.
+   As usual, Datastore would ignore artifacts outside of its root instead of deleting them.
+
+Once again, we've eliminated any alternation between database/server calls and Datastore operations, reducing latency.
+We've also avoided any database transactions over datastore operations.
+
+
+``Butler.purge``
+""""""""""""""""
+
+This is another new proposed interface, this time for fully removing multiple datasets from both Registry and Datastore.
+It's the rest of the replacement for ``Butler.pruneDatasets`` and reimplementation for ``Butler.removeRuns``.
+
+The approach here is essentially identical to ``Butler.unstore``, but in the first step we would instruct the Registry to remove all references to the datasets, not just the datastore records.
+
+.. _adding_journal_files:
+
+Adding journal files
+--------------------
+
+The main flaw in the proposal above is that it can leave artifacts in the Datastore root that are untracked and hard to find, due to both I/O failures and abandoned batch runs.
```

Comment:
batch also has the problem of not writing an output file when we expected one. This is why transfer_from has to check for the existence of 250k files in an object store and is the main reason it's slow. We talked about having a per-quantum look up file which would describe the actual files that were written and that would speed things up.

Reply:
I completely agree, and would still like to go in that direction someday.  But I think it's not a consistency problem so I wasn't going to get into it here.